### DISCUSSION TITLE
Passing `transport_options` to `asyncssh` driver

### DISCUSSION THREAD

**Discussion by cuong-nguyenduy** [carlmontanari/scrapli#175] [Q&A] [answered]
Hi,

I need to use `asyncssh` as the driver to access to one of our Huawei device. The device is running an old version which only supports legacy algorithms.

Is there anyway to inject the options, especially the inclusion of legacy algorithms? I tried with `transport_options` as below, it doesn't seem to work.

I wanted to try with this code:
```python
import asyncio
from scrapli import AsyncScrapli

from private import *  # including username and password


async def worker(ip):
    driver_options = {
        "host": ip,
        "auth_username": username,
        "auth_password": password,
        "auth_strict_key": False,
        "transport": "asyncssh",
        "platform": "huawei_vrp",
        "transport_options": {
            "kex_algs": ["diffie-hellman-group14-sha1", "diffie-hellman-group1-sha1", "diffie-hellman-group-exchange-sha1"],
            "encryption_algs": ["aes256-cbc", "aes192-cbc", "aes256-ctr", "aes192-ctr", "aes128-ctr"],
        },
    }

    async with AsyncScrapli(**driver_options) as conn:
        response = await conn.get_prompt()
        print("Prompt is", response)


async def run(loop, ip):
    task = loop.create_task(
        worker(ip),
    )
    while not task.done():
        await asyncio.sleep(1)

loop = asyncio.get_event_loop()
loop.run_until_complete(run(loop, "172.30.0.19"))
```

I received Exception when running the script.
```bash
Task exception was never retrieved
future: <Task finished name='Task-2' coro=<worker() done, defined at /Users/nguyendc/Projects/remote_debug/asyncssh_options.py:7> exception=KeyExchangeFailed('No matching key exchange algorithm found, sent curve25519-sha256,curve25519-sha256@libssh.org,curve448-sha512,ecdh-sha2-nistp521,ecdh-sha2-nistp384,ecdh-sha2-nistp256,ecdh-sha2-1.3.132.0.10,diffie-hellman-group-exchange-sha256,diffie-hellman-group14-sha256,diffie-hellman-group15-sha512,diffie-hellman-group16-sha512,diffie-hellman-group17-sha512,diffie-hellman-group18-sha512,diffie-hellman-group14-sha256@ssh.com,diffie-hellman-group14-sha1,rsa2048-sha256 and received diffie-hellman-group-exchange-sha1,diffie-hellman-group1-sha1')>
Traceback (most recent call last):
  File "/Users/nguyendc/Projects/remote_debug/asyncssh_options.py", line 21, in worker
    async with AsyncScrapli(**driver_options) as conn:
  File "/Users/nguyendc/.pyenv/versions/3.9.5-dev/lib/python3.9/site-packages/scrapli/driver/base/async_driver.py", line 40, in __aenter__
    await self.open()
  File "/Users/nguyendc/.pyenv/versions/3.9.5-dev/lib/python3.9/site-packages/scrapli/driver/base/async_driver.py", line 82, in open
    await self.transport.open()
  File "/Users/nguyendc/.pyenv/versions/3.9.5-dev/lib/python3.9/site-packages/scrapli/transport/plugins/asyncssh/transport.py", line 114, in open
    self.session = await asyncio.wait_for(
  File "/Users/nguyendc/.pyenv/versions/3.9.5/lib/python3.9/asyncio/tasks.py", line 481, in wait_for
    return fut.result()
  File "/Users/nguyendc/.pyenv/versions/3.9.5/lib/python3.9/asyncio/tasks.py", line 690, in _wrap_awaitable
    return (yield from awaitable.__await__())
  File "/Users/nguyendc/.pyenv/versions/3.9.5-dev/lib/python3.9/site-packages/asyncssh/connection.py", line 6803, in connect
    return await _connect(options, loop, flags, conn_factory,
  File "/Users/nguyendc/.pyenv/versions/3.9.5-dev/lib/python3.9/site-packages/asyncssh/connection.py", line 303, in _connect
    await conn.wait_established()
  File "/Users/nguyendc/.pyenv/versions/3.9.5-dev/lib/python3.9/site-packages/asyncssh/connection.py", line 2243, in wait_established
    await self._waiter
  File "/Users/nguyendc/.pyenv/versions/3.9.5-dev/lib/python3.9/site-packages/asyncssh/connection.py", line 1047, in data_received
    while self._inpbuf and self._recv_handler():
  File "/Users/nguyendc/.pyenv/versions/3.9.5-dev/lib/python3.9/site-packages/asyncssh/connection.py", line 1287, in _recv_packet
    processed = handler.process_packet(pkttype, seq, packet)
  File "/Users/nguyendc/.pyenv/versions/3.9.5-dev/lib/python3.9/site-packages/asyncssh/packet.py", line 215, in process_packet
    self._packet_handlers[pkttype](self, pkttype, pktid, packet)
  File "/Users/nguyendc/.pyenv/versions/3.9.5-dev/lib/python3.9/site-packages/asyncssh/connection.py", line 1866, in _process_kexinit
    kex_alg = self._choose_alg('key exchange', kex_algs, peer_kex_algs)
  File "/Users/nguyendc/.pyenv/versions/3.9.5-dev/lib/python3.9/site-packages/asyncssh/connection.py", line 1119, in _choose_alg
    raise KeyExchangeFailed(
asyncssh.misc.KeyExchangeFailed: No matching key exchange algorithm found, sent curve25519-sha256,curve25519-sha256@libssh.org,curve448-sha512,ecdh-sha2-nistp521,ecdh-sha2-nistp384,ecdh-sha2-nistp256,ecdh-sha2-1.3.132.0.10,diffie-hellman-group-exchange-sha256,diffie-hellman-group14-sha256,diffie-hellman-group15-sha512,diffie-hellman-group16-sha512,diffie-hellman-group17-sha512,diffie-hellman-group18-sha512,diffie-hellman-group14-sha256@ssh.com,diffie-hellman-group14-sha1,rsa2048-sha256 and received diffie-hellman-group-exchange-sha1,diffie-hellman-group1-sha1
```

The device allows ssh into it, with the options as below:
```bash
Host 172.*
	KexAlgorithms +diffie-hellman-group1-sha1
	KexAlgorithms +diffie-hellman-group14-sha1
	KexAlgorithms +diffie-hellman-group-exchange-sha1
	Ciphers +aes256-cbc
	Ciphers +aes192-cbc
	Ciphers +aes256-ctr
	Ciphers +aes192-ctr
	Ciphers +aes128-ctr
```

**Comment by carlmontanari** [comment]
Hi @cuong-nguyenduy 

I am 99.9% sure that asyncssh will honor the kex/cipher/etc options in your ssh config file (see also #173 for a related discussion that happened recently). So you should be able to just have your ssh config setup as you outline and add `ssh_config_file=True` (or pass a path to your config file) which will make scrapli pass that along to asyncssh where it will take care of the parsing/setting of options.

Let us know if that gets you sorted!

Carl

**Reply by cuong-nguyenduy** [reply]
Hi @carlmontanari 

Indeed, with option `"ssh_config_file": True`, the script now works as expected.

But this isn't what I intend to. I want to ignore the legacy on first try, and enable them if the `KeyExchangeFailed` exception is raised. In my PC, I have the ability to change `ssh_config`; on the server it is not possible though.

I also dig into the code and realised somehow `transport_options` isn't carried all the way to `assyncssh`. I may be mistakenly to think that this parameter `transport_options` is for us to carry such options to the transport driver. Am I right?

Anyway, when changing this file `scrapli/transport/plugins/asyncssh/transport.py` as below, I can pass the `transport_options` and achieve the same result **without setting** `"ssh_config_file": True`

```diff
@@ -109,6 +109,9 @@ class AsyncsshTransport(AsyncTransport):
             "agent_path": None,
             "config": self.plugin_transport_args.ssh_config_file,
         }
+
+        if hasattr(self._base_transport_args, "transport_options"):
+            common_args.update(self._base_transport_args.transport_options)

         try:
             self.session = await asyncio.wait_for(
```

I'm wondering if this is the right way, though.
@carlmontanari do you have any comment? Is there a better way to do this?

Cuong.

**Reply by carlmontanari** [reply]
Ok I think I gotchya.

"Yes" `transport_options` is meant for stuff like this, but so far (until now it seems!) there has not been any need for using any options in asyncssh -- mostly the `transport_options` have been used for the `system` transport like you can see [here](https://github.com/carlmontanari/scrapli/blob/develop/scrapli/transport/plugins/system/transport.py#L34-L43). 

The `transport_options` *are* being set for asyncssh (in the base class [here-ish as part of the base transport args](https://github.com/carlmontanari/scrapli/blob/develop/scrapli/transport/base/base_transport.py#L11-L26)) they just are currently ignored. 

As this also seems relevant to #173 I'd be open to a PR that adds support for explicitly passing cipher/kext/whatever via transport args to asyncssh if you were up for taking a crack at it?

Carl

**Reply by cuong-nguyenduy** [reply]
@carlmontanari 

I am willing to create a PR. In fact, I have my commit here - https://github.com/cuong-nguyenduy/scrapli/commit/e95a4d6f0133d41518eec91aa07476adf119f3fc

However, I'm not sure how to add the tests, though. It is quite overwhelmed for me right now. Let me see what I can do.

Cuong.

**Reply by carlmontanari** [reply]
Sounds great @cuong-nguyenduy -- happy to help work through the PR with ya and help with tests and anything! I don't think you should need that `hasattr` btw -- the transport options should always be set on the base transport args.

**Comment by cuong-nguyenduy** [comment, marked as answer]
@carlmontanari
I created a PR https://github.com/carlmontanari/scrapli/pull/178 for this. I ran the test in my lab with a few devices (due to my limited ability to run the actual tests).

Please let me know if it's ok or needs improvement.

Cuong.
